### DISCUSSION TITLE
SysTray: Checking which backend is in use

### DISCUSSION THREAD

**Discussion by elParaguayo** [qtile/qtile#2523] [Q&A] [answered]
@m-col Another one for you...

Given your comment about `SysTray` not working on wayland, I had hoped to make the widget return a text box if the backend is not x11. A neat way to do this would be to add a `__new__` method to `SysTray` and have it return a `ConfigErrorWidget` instance with a message saying that `SysTray` in unavailable. We can also add a comment about x11 only to the widget's docstring but I'm guessing some people won't read the docs!

The issue with this approach is that  I don't seem to be able to access `qtile.core.name` when the widget is being created. Is that because the core is not running at the time the config file is being parsed?

I could have made a mistake somewhere but it's definitely that line that's failing. If I log the output of `dir(qtile)`, `core` is not listed as an attribute.

**Comment by m-col** [comment, marked as answer]
I've been thinking about having some kind of mechanism for widgets to specify
which backends they are limited to, if any, too. My most recent idea was to have
a `BackendNotSupportedError` or something that is raised during
`widget._configure` if `qtile.core.name` isn't in that widget's supported
backend list. Then `Bar` can catch it and replace it with a `ConfigErrorWidget`
as you suggest, like what currently happens when a widget crashes during
`_configure`. `qtile` is passed to `_configure` so `qtile.core.name` should be
accessible just fine there. Although, `qtile.core.name` is also accessible
during config reading so not sure what's happening when you're seeing that :)

On 210609 08:33, elParaguayo wrote:
***@***.*** Another one for you...

**Reply by elParaguayo** [reply]
Nice. I like your approach better!

**Comment by m-col** [comment]
The reason I hadn't implemented this is already is that we might not actually
need to do it. If we can implement a systray equivalent for wayland using the
non-x11 specific spec then `widget.Systray` can use that under Wayland.

Other widgets that need(ed) updating to be backend-agnostic are:

  - KeyboardLayout (Wayland support just been merged)
  - TaskList - might just be the window icons that need supporting
  - Prompt (simple-ish, need to forward key events to Bar)
  - Clipboard (need to fix Wayland clipboard first)
  - CapsNumLockIndicator

**Comment by elParaguayo** [comment]
I had a quick look at the SystemNotifier idea. It looks interesting. Should be doable with dbus-next but may take a little bit of time.
